### ISSUE TITLE
Ctrl+. -> generate field puts generated field inside #region IDisposable Support

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#37401]
Supposecode snippet 1 below. Place cursor on the constructor's parameter, press Ctrl+. and choose any of the &quot;Create and initialize...&quot; menu options. The generated code will be placed inside the &quot;#region IDisposable support&quot;, like code snippet number two.

I would expect it to be generated in front of the constructor.

**Code snippet 1**
Does not declare or initialize afield or property to hold the constructor parameter.
```
public class Disposable : IDisposable
	{
		public Disposable(string something)
		{
		}

#region IDisposable Support
		private bool disposedValue = false; // To detect redundant calls

protected virtual void Dispose(bool disposing)
		{
		}
		public void Dispose()
		{
			Dispose(true);
		}
		#endregion
	}
```

**Code snippet 2**
After using the &quot;Create and initialize field 'something'&quot; code completion tool. Note the generated field is placed on line 9 inside the IDisposable region.
```
public class Disposable : IDisposable
	{
		public Disposable(string something)
		{
			this.something = something;
		}

#region IDisposable Support
		private bool disposedValue = false; // To detect redundant calls
		private readonly string something;

protected virtual void Dispose(bool disposing)
		{
		}
		public void Dispose()
		{
			Dispose(true);
		}
		#endregion
	}
```

_This issue has been moved from https://developercommunity.visualstudio.com/content/problem/652913/ctrl-generate-field-puts-generated-field-inside-re.html
VSTS ticketId: 949988_
_These are the original issue comments:_

Visual Studio Feedback System on 7/22/2019, 01:35 AM (6 hours ago): <markdown><p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>
</markdown>
_These are the original issue solutions:_
(no solutions)

**Comment by CyrusNajmabadi** [comment]
I'm not certain there's anything reasonable that can be done here. For all we know, the user **wants** the fields in that region. For example, the region might have been `#region state, in which case, not putting it there with the other fields would be incorrect.

**Comment by svethors** [comment]
As the user not , I would just expect it to work consistently.

Case with no region:
`public class Test
	{
		public Test(int myInt)
		{
		}
	}
`
Result:
`public class Test
	{
		private readonly int myInt;

		public Test(int myInt)
		{
			this.myInt = myInt;
		}
	}
`

Result with a region (generated by the implement with IDisposable pattern:
`public class Test : IDisposable
	{
		public Test(int myInt)
		{
			this.myInt = myInt;
		}

		#region IDisposable Support
		private bool disposedValue = false; // To detect redundant calls
		private readonly int myInt;

		protected virtual void Dispose(bool disposing)
		{
			if (!disposedValue)
			{
				if (disposing)
				{
					// TODO: dispose managed state (managed objects).
				}

				// TODO: free unmanaged resources (unmanaged objects) and override a finalizer below.
				// TODO: set large fields to null.

				disposedValue = true;
			}
		}

		// TODO: override a finalizer only if Dispose(bool disposing) above has code to free unmanaged resources.
		// ~Test()
		// {
		//   // Do not change this code. Put cleanup code in Dispose(bool disposing) above.
		//   Dispose(false);
		// }

		// This code added to correctly implement the disposable pattern.
		public void Dispose()
		{
			// Do not change this code. Put cleanup code in Dispose(bool disposing) above.
			Dispose(true);
			// TODO: uncomment the following line if the finalizer is overridden above.
			// GC.SuppressFinalize(this);
		}
		#endregion
	}`

**Comment by svethors** [comment]
If I have a field before the constructor (with no region), and a field inside a region below the constructor, the generated field will end up inside the region below:

```
public class Test
	{
		int someExistingField;
		public Test(int myInt)
		{
			this.myInt = myInt;
		}
		#region bottom
		int someExistingField2;
		private readonly int myInt;
		#endregion
	}
```
If I have a region both before and after the constructor:

If there are no existing fields, I will get the generated field on top and before any region:
```
public class Test
	{
		private readonly int myInt;
		#region top
		#endregion

		public Test(int myInt)
		{
			this.myInt = myInt;
		}

		#region bottom
		#endregion
	}
``` 
The same if I add an existing field to the top region:
``` 
public class Test
	{
		private readonly int myInt;
		#region top
		int someExistingField;
		#endregion

		public Test(int myInt)
		{
			this.myInt = myInt;
		}

		#region bottom
		#endregion
	}
``` 
But if I add an existing field to the bottom region, suddenly the generated field is **inside** a region rather than outside, and the region chosen is the bottom one:
``` 
public class Test
	{
		#region top
		int someExistingField;
		#endregion

		public Test(int myInt)
		{
			this.myInt = myInt;
		}

		#region bottom
		int someExistingField2;
		private readonly int myInt;
		#endregion
	}
``` 
And this is what happens even if I add other fields outside and inside regions, above and below. It seems as if the rule is: "If there is a region with a field in it at the bottom of a class, add generated field **inside** the bottom region. If not, add generated code **outside** any region and **above** any top region".

**Comment by CyrusNajmabadi** [comment]
I believe the rule is: add the last group of fields we find.  Otherwise, add at the top of the class.

> As the user reporting it, I would just expect it to work consistently.

I believe it is consistent.  If you have fields, we assume you placed them where you want fields to be placed.  So we follow that.  Absent any fields we have to guess where you want them, and we follow the very normal pattern of having them at teh top of the class.

**Comment by CyrusNajmabadi** [comment]
> If I have a field before the constructor (with no region), and a field inside a region below the constructor, the generated field will end up inside the region below:

Yes, this follows the rule of placing fields with existing fields if they exist.

> If there are no existing fields, I will get the generated field on top and before any region:

yes, this follows the rule of adding to the top of the class if there are no fields.

> The same if I add an existing field to the top region:

Yes.  It was added to the same group.  I believe we tried to keep them alphabetically sorted here since htere wasn't much else to go on.  So it was placed before.

> And this is what happens even if I add other fields outside and inside regions, above and below

Yes, i believe this goes into the category of "place into the last group of fields".

**Comment by CyrusNajmabadi** [comment]
Main logic is here:

https://github.com/dotnet/roslyn/blob/dfc664f7a9eba2d9c6d09a4e3bb8910661d6086f/src/Workspaces/CSharp/Portable/CodeGeneration/FieldGenerator.cs#L57-L58

It says: place after the last field if we have one, otherwise before the first member.  You can see the rest of hte logic for how it is placing things in that file.

**Comment by CyrusNajmabadi** [comment]
General idea:

if we're generating something const, we try to keep together with your other consts.  Otherwise, consts go at the top of the class.

If you're generating something readonly/normal, we attempt to place after the last existing field that matches it.  Otherwise, we'll place with the rest of the fields.  And finally we'll place with the rest of the consts.  Absent any of that, we place at the top of the class.

**Comment by svethors** [comment]
I see what you're saying. So then it is the "implement with IDisposable pattern" snippet that ends up making this very nice feature quite annoying :)

If I add one private readonly field at the top of the class, the generated field ends up there too. So doing generate field for parameter, and leaving in readonly, before doing the implement IDisposable, works as a sort of workaround.

Though if I could wish, I would be happy with an optional "always generate fields and properties above the constructor"-rule.

**Comment by CyrusNajmabadi** [comment]
> Though if I could wish, I would be happy with an optional "always generate fields and properties above the constructor"-rule.

Indeed.  That's something we've noodled on for years.  There's no easy and obvious way to give users the flexibility that may be desired here, so it's been on hte backburner forever.

**Comment by svethors** [comment]
Or, could the "implement with IDisposable pattern" be done in such a way that it too added the one "problematic" field in accordance with the generate field rule set?

**Comment by CyrusNajmabadi** [comment]
@svethors Definitely.  I frankly don't know why it uses that pattern in the first place.  We moved off of regions ages ago in our default code-spit.

**Comment by svethors** [comment]
As a side note if the IDisposable generation is changed: If all the TODO:'s could go away and just leave the comments, that would also get rid of loads of unnecessary cluttering warnings.

**Comment by CyrusNajmabadi** [comment]
> Or, could the "implement with IDisposable pattern" be done in such a way that it too added the one "problematic" field in accordance with the generate field rule set?

This is now how we generate the pattern.  There are no regions, and fields are properly grouped with other fields.  Closing htis out.
